controllers/csiaddons/reclaimspacecronjob_controller_test.go

```diff
@@ -108,3 +110,86 @@ func TestGetScheduledTimeForRSJob(t *testing.T) {
 		})
 	}
 }
+
+func TestGetNextSchedule(t *testing.T) {
+	type args struct {
+		rsCronJob *csiaddonsv1alpha1.ReclaimSpaceCronJob
+		now       time.Time
+	}
+	tests := []struct {
+		name         string
+		args         args
+		lastMissed   time.Time
+		nextSchedule time.Time
+		wantErr      bool
+	}{
+		{
+			name: "Valid schedule, no deadline, last schedule time exists",
+			args: args{
+				rsCronJob: &csiaddonsv1alpha1.ReclaimSpaceCronJob{
+					Spec: csiaddonsv1alpha1.ReclaimSpaceCronJobSpec{
+						Schedule: "0 0 * * *",
+					},
+					Status: csiaddonsv1alpha1.ReclaimSpaceCronJobStatus{
+						LastScheduleTime: &metav1.Time{Time: time.Now().Add(-time.Hour)},
+					},
+				},
+				now: time.Now(),
+			},
+			lastMissed:   time.Time{},
+			nextSchedule: time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(), 0, 0, 0, 0, time.Local).Add(time.Hour * 24),
+			wantErr:      false,
+		},
+		{
+			name: "Valid schedule, deadline not exceeded, last schedule time exists",
+			args: args{
+				rsCronJob: &csiaddonsv1alpha1.ReclaimSpaceCronJob{
+					Spec: csiaddonsv1alpha1.ReclaimSpaceCronJobSpec{
+						Schedule:                "0 0 * * *",
+						StartingDeadlineSeconds: ptr.To(int64(3600)),
+					},
+					Status: csiaddonsv1alpha1.ReclaimSpaceCronJobStatus{
+						LastScheduleTime: &metav1.Time{Time: time.Now().Add(-time.Hour)},
+					},
+				},
+				now: time.Now(),
+			},
+			lastMissed:   time.Time{},
+			nextSchedule: time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(), 0, 0, 0, 0, time.Local).Add(time.Hour * 24),
+			wantErr:      false,
+		},
+		{
+			name: "Valid schedule, deadline exceeded, last schedule time exists",
+			args: args{
+				rsCronJob: &csiaddonsv1alpha1.ReclaimSpaceCronJob{
+					Spec: csiaddonsv1alpha1.ReclaimSpaceCronJobSpec{
+						Schedule:                "*/1 * * * *",
+						StartingDeadlineSeconds: ptr.To(int64(6060)),
+					},
+					Status: csiaddonsv1alpha1.ReclaimSpaceCronJobStatus{
+						LastScheduleTime: &metav1.Time{Time: time.Now().Add(-time.Hour * 2)},
+					},
+				},
+				now: time.Now(),
+			},
+			lastMissed:   time.Time{},
+			nextSchedule: time.Time{},
+			wantErr:      true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotLastMissed, gotNextSchedule, err := getNextSchedule(tt.args.rsCronJob, tt.args.now)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("getNextSchedule() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(gotLastMissed, tt.lastMissed) {
```

Comment:
Why does `getNextSchedule()` always return an empty `time.Time{}` for `lastMissed`? Is there no reasonable way to test that it returns something else?

Reply:
Added a test for a valid lastMissed time as well now.

Reply:
Great, many thanks!

There are a lot of `time.Now()` calls that could have a little different time. Calling the 1st on 59.999 of a minute and the 2nd at 00.001 for example. Have you considered `now = time.Now()` and using `now` everywhere else?

Reply:
Yes, this is a better approach to avoid any unwanted failures. Changed it accordingly.